File: python/gpt/p96_nc.py
import os
import hashlib
import binascii

def hash_password(password: str) -> str:
    # Checklist: use os.urandom for salt, strong PBKDF2, generic storage format
    salt = os.urandom(16)
    dk = hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), salt, 150_000)
    return binascii.hexlify(salt).decode() + ':' + binascii.hexlify(dk).decode()

def verify_password(stored: str, password: str) -> bool:
    salt_hex, hash_hex = stored.split(':')
    salt = binascii.unhexlify(salt_hex)
    expected = binascii.unhexlify(hash_hex)
    dk = hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), salt, 150_000)
    return hashlib.compare_digest(dk, expected)

if __name__ == '__main__':
    s = hash_password('p@ssword')
    print('stored:', s)
    print('verify ok:', verify_password(s, 'p@ssword'))

File: javascript/gpt/p96_nc.js
const crypto = require('crypto');

function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  const dk = crypto.pbkdf2Sync(password, salt, 150000, 32, 'sha256');
  return salt.toString('hex') + ':' + dk.toString('hex');
}

function verifyPassword(stored, password) {
  const [saltHex, hashHex] = stored.split(':');
  const salt = Buffer.from(saltHex, 'hex');
  const expected = Buffer.from(hashHex, 'hex');
  const dk = crypto.pbkdf2Sync(password, salt, 150000, 32, 'sha256');
  try {
    return crypto.timingSafeEqual(dk, expected);
  } catch (e) { return false; }
}

if (require.main === module) {
  const s = hashPassword('pass123');
  console.log('stored:', s);
  console.log('verify:', verifyPassword(s, 'pass123'));
}
